edu_edfi_airflow/dags/edfi_resource_dag.py

```diff
@@ -1,42 +1,57 @@
 from airflow.operators.python import PythonOperator
 from airflow.utils.task_group import TaskGroup
 
-from ea_airflow_util import slack_callbacks, update_variable
+from ea_airflow_util import EACustomDAG
+from ea_airflow_util import update_variable
 from edfi_api_client import camel_to_snake
 
-from edu_edfi_airflow.dags.callables import change_version
-from edu_edfi_airflow.dags.dag_util import airflow_util
-from edu_edfi_airflow.providers.edfi.transfers.edfi_to_s3 import EdFiToS3Operator
-from edu_edfi_airflow.providers.snowflake.transfers.s3_to_snowflake import S3ToSnowflakeOperator
+from edu_edfi_airflow.callables import airflow_util, change_version
+from edu_edfi_airflow.providers.edfi.transfers.edfi_to_s3 import EdFiToS3Operator, BulkEdFiToS3Operator
+from edu_edfi_airflow.providers.snowflake.transfers.s3_to_snowflake import BulkS3ToSnowflakeOperator
 
 
 class EdFiResourceDAG:
     """
-
+    If use_change_version is True, initialize a change group that retrieves the latest Ed-Fi change version.
+    If full_refresh is triggered in DAG configs, reset change versions for the resources being processed in Snowflake.
+
+    DAG Structure:
+        (Ed-Fi3 Change Version Window) >> [Ed-Fi Resources/Descriptors (Deletes/KeyChanges)] >> (increment_dbt_variable) >> dag_state_sentinel
+    
+    "Ed-Fi3 Change Version Window" TaskGroup:
+        get_latest_edfi_change_version >> reset_previous_change_versions_in_snowflake
+
+    "Ed-Fi Resources/Descriptors (Deletes/KeyChanges)" TaskGroup:
+        (get_cv_operator) >> [Ed-Fi Endpoint Task] >> copy_all_endpoints_into_snowflake >> (update_change_versions_in_snowflake)
+
+    There are three types of Ed-Fi Endpoint Tasks. All take the same inputs and return the same outputs (i.e., polymorphism).
+        "Default" TaskGroup
+        - Create one task per endpoint.
+        "Dynamic" TaskGroup
+        - Dynamically task-map all endpoints. This function presumes that only endpoints with deltas to ingest are passed as input.
+        "Bulk" TaskGroup
+        - Loop over each endpoint in a single task.
+
+    All task groups receive a list of (endpoint, last_change_version) tuples as input.
+    All that successfully retrieve records are passed onward as a (endpoint, filename) tuples to the S3ToSnowflake and UpdateSnowflakeCV operators.
     """
-    newest_edfi_cv_task_id = "get_latest_edfi_change_version"  # Original name for historic run compatibility
-    previous_snowflake_cv_task_id = "get_previous_change_versions_from_snowflake"
-
-    params_dict = {
-        "full_refresh": Param(
-            default=False,
-            type="boolean",
-            description="If true, deletes endpoint data in Snowflake before ingestion"
-        ),
-        "endpoints": Param(
-            default=[],
-            type="array",
-            description="Newline-separated list of specific endpoints to ingest (case-agnostic)\n(Bug: even if unused, enter a newline)"
-        ),
+    DEFAULT_CONFIGS = {
```

Comment:
A major structural change of this DAG is to internalize endpoint config bundle structure. These are referenced throughout the polymorphic task groups, so consistency is required. Method `build_endpoint_configs()` takes the user-provided configs and adds missing defaults.